### PR TITLE
Feature/py open ssl

### PR DESCRIPTION
**Pull request by SvenRoederer**
This is my idea to issue #14  - this is WIP and only for discussion, but generated certificates match currently used 
- uses PyOpenSSL to do the key-generation and Cert-signing 
- no use of "easyrsa" shell calls
- uniqueness of cert serial is assured by database 

please comment on this

### CODE REVIEW

**Comment by SvenRoederer** [comment]
is this idea to bad, that nobody likes to check on this?

**Comment by booo** [comment]
We should nice up the code a bit but the direction is perfect!

I suggest that we generate the id for the request in the database. Otherwise we may run into issues with multiple process creating the same id. If we generate the id when the request enters the database we may waste some id space when we reject the request but who cares?

File: buildcert.py [hunk 1]
```diff
@@ -7,16 +7,82 @@
...
+
+# taken from https://gist.github.com/ril3y/1165038
+def create_cert(cert_name, cert_email, cert_sn, cert_key):
+    """
+    If datacard.crt and datacard.key don't exist in cert_dir, create a new
+    self-signed cert and keypair and write them into that directory.
+    """
+
+    # get required CA-data
+    ca_cert_file = open('/tmp/ffca.crt', 'r')
```
Comment (booo) [inline buildcert.py]: This should be part of config.py

File: buildcert.py [hunk 2]
```diff
@@ -7,16 +7,82 @@
...
+    ca_cert = crypto.load_certificate(crypto.FILETYPE_PEM, ca_cert_file.read())
+    ca_key_file = open('/tmp/ffca.key', 'r')
+    ca_key = crypto.load_privatekey(crypto.FILETYPE_PEM, ca_key_file.read())
+
+    if True:
+
+        # create a self-signed cert
+        cert = crypto.X509()
+        cert.set_version(0x02)  # X509-Version 3
+        cert.get_subject().C = "DE"
```
Comment (booo) [inline buildcert.py]: This should be part of config.py too.

File: buildcert.py [hunk 3]
```diff
@@ -7,16 +7,82 @@
...
+        cert.get_subject().ST = "Eastern Germany"
+        cert.get_subject().L = "Berlin"
+        cert.get_subject().O = "Foerderverein Freie Netzwerke e.V."
+        cert.get_subject().CN = "freifunk_%s" % cert_name
+        cert.get_subject().emailAddress = cert_email
+        cert.set_serial_number(cert_sn)
+        cert.gmtime_adj_notBefore(0)
+        cert.gmtime_adj_notAfter(10*365*24*60*60)
```
Comment (booo) [inline buildcert.py]: Same here is true for notAfter

File: buildcert.py [hunk 4]
```diff
@@ -7,16 +7,82 @@
...
+            crypto.X509Extension(b'basicConstraints', False, b"CA:FALSE"),
+            crypto.X509Extension(b'nsComment', False, b'made for you with PyOpenSSL'),
+            crypto.X509Extension(b'subjectKeyIdentifier', False, b'hash', subject=cert),
+            crypto.X509Extension(b'authorityKeyIdentifier', False, b'keyid:always,issuer:always', issuer=ca_cert),
+            crypto.X509Extension(b'extendedKeyUsage', False, b'TLS Web Client Authentication'),
+            crypto.X509Extension(b'keyUsage', False, b'Digital Signature'),
+        ]
+        cert.add_extensions(cert_ext)
+
+        cert.sign(ca_key, 'sha1')
```
Comment (booo) [inline buildcert.py]: Can we just switch to sha2? If we want to use the certificates in the future that's maybe a "good" move.

File: buildcert.py [hunk 5]
```diff
@@ -7,16 +7,82 @@
...
+        print('generating key')
+        new_key = create_key()
         print('generating certificate')
-        call([app.config['COMMAND_BUILD'], request.id, request.email])
-        call([app.config['COMMAND_MAIL'], request.id, request.email])
-        request.generation_date = datetime.date.today()
+        new_cert_sn = Request.getMaxCertSn() + 1
+        request.cert_sn = new_cert_sn
+        new_cert = create_cert(request.id, request.email, request.cert_sn, new_key)
+        print (crypto.dump_certificate(crypto.FILETYPE_TEXT, new_cert))
```
Comment (booo) [inline buildcert.py]: We should save the cert/key somewhere. I guess openvpn expects the keys at least for revocation to be somewhere.

**Comment by booo** [comment]
If you move this to a branch inside the repository we can collaborate on this.

Thanks for the proposal. Sorry for the late replay here.

**Comment by SvenRoederer** [comment]
- I agree that some setting have to move to config.py
- Cert-serial: check commit [model] add class-method "getMaxCertSn()" --> S/N is generated by DB or will at least cause a constraint violation. current "database-id" is the is-token entered on web-site. but per requirement the Certs-ID has to be numerical and of course unique.
- shall we store the keys and certs in the DB or in file-archives as we do now?
- for revocation we have to think of a CRL-infrastructure ... and the CRLs only need the final cert or just the SN.

**Comment by booo** [comment]
There is a revocation script. Can you check what it does and  what openvpn expects? Maybe this will answer the question of where to store the keys/certs too.

**Comment by SvenRoederer** [comment]
for revocation OpenVPN expects the regular CRLs as file on disk.
Genrating the CRLs can be done offline via openssl command or even integrated into the web-app (for downloading by cron on each VPN-server)

**Comment by SvenRoederer** [comment]
uploaded an improved version (including use of config.py) to branch https://github.com/freifunk-berlin/ca.berlin.freifunk.net/tree/pyOpenSSL 

closing this RFC-PR for further work on [ 3505097 ]

**Comment by SvenRoederer** [comment]
@booo @KenTacki - please comment on this
- needs some cleanup + rebasing
- seems to work for me
- generates "dropin" certificates / as much compatibility as possible to current setup

**Comment by SvenRoederer** [comment]
looks like it was auto-reopened

**Pull request closed without merging**